### ISSUE TITLE
Refine lints about field and method removals to account for `Deref/DerefMut`

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#766]
This issue about trait associated consts and functions acting as a backstop for removed inherent items got me thinking, what about Deref? Is the following breaking if the `// REMOVE` lines are removed?

```rust
pub struct Foo {
    pub field: (), // REMOVE
    inner: Inner,
}

impl Foo {
    pub fn method(&self) {} // REMOVE

    pub fn new() -> Self {
        Self {
            field: (), // REMOVE
            inner: Inner { field: () },
        }
    }
}

pub struct Inner {
    pub field: (),
}

impl Inner {
    pub fn method(&self) {}
}

impl core::ops::Deref for Foo {
    type Target = Inner;
    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl core::ops::DerefMut for Foo {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.inner
    }
}

fn test() {
    let mut foo = Foo::new();

    foo.field = ();
    foo.method();
}
```

_Originally posted by @jw013 in https://github.com/obi1kenobi/cargo-semver-checks/issues/727#issuecomment-2065620946_

The long-term solution we want here is similar to the case of [`clippy::correctness` vs `clippy::suspicious`](https://doc.rust-lang.org/clippy/lints.html):
- We want a `correctness` lint (on by default, error level) that allows `Deref/DerefMut` to be used to replace the removed item.
- We want a `suspicious` lint (on by default, warning level) that does not allow the above, and never fires if the above lint is already going to fire.
